### ISSUE TITLE
ListObjectsV2 fails to list empty folders

### ISSUE THREAD

**Issue by chrisharward** [minio/minio#11370]
If I create an empty folder in Amazon S3 (a zero byte object with a trailing slash in its name), and then run a ListObjects_v2 using the object key as the prefix of the operation, Amazon returns the empty folder in the ListBucketResult. I find that doing the same on MinIO does not.

## Expected Behavior
If I create a zero byte object with a trailing slash in its key (say "prefix/"), and then run a ListObjects_v2 request with prefix=prefix/, I expect the ListBucketResult to have a single entry in the Contents field for the prefix/ item.

## Current Behavior
The Contents member of the ListBucketResult is empty.

## Possible Solution
Not sure.

## Context
We use MinIO to implement system/integration tests for a system aimed at running against Amazon S3, so having it match the behavior of Amazon S3 as closely as possible makes that more effective and reduces the amount of testing we need to do against Amazon S3.

## Regression
Not that I know of.

## Your Environment
This is reproduceable in play.minio.io

**Comment by harshavardhana** [comment]
@chrisharward please share a sample code.

**Comment by chrisharward** [comment]
Something like this:

`import io

from minio import Minio
from minio.error import S3Error

def main():
    # Create a client with the MinIO server playground, its access key
    # and secret key.
    client = Minio(
        "play.min.io",
        access_key="Q3AM3UQ867SPQQA43P2F",
        secret_key="zuf+tfteSlswRu7BJ86wekitnifILbZam1KYY3TG",
    )

    # Upload empty folder
    result = client.put_object(
        "test", "prefix/", io.BytesIO(b""), 0,
    )
    print("created {0} object; etag: {1}, version-id: {2}".format(
            result.object_name, result.etag, result.version_id,
        ),
    )

    objects = client.list_objects("test", prefix="prefix/", recursive=True)
    count = 0
    for obj in objects:
        count += 1
        print(obj)
    print("{} objects at prefix {}".format(count, "prefix/"))

if __name__ == "__main__":
    try:
        main()
    except S3Error as exc:
        print("error occurred.", exc)`

**Comment by vadmeste** [comment]
The issue is that the code thought it wants to list a directory but in this case it should list with a prefix `__XLDIR__`

Do you think the fix below is good enough @harshavardhana, @klauspost ?


```diff
diff --git a/cmd/metacache-walk.go b/cmd/metacache-walk.go
index e9130e38e..bda3a3470 100644
--- a/cmd/metacache-walk.go
+++ b/cmd/metacache-walk.go
@@ -96,6 +96,12 @@ func (s *xlStorage) WalkDir(ctx context.Context, opts WalkDirOptions, wr io.Writ
        var scanDir func(path string) error
        scanDir = func(current string) error {
                entries, err := s.ListDir(ctx, opts.Bucket, current, -1)
+               // Retry listing if the path has a leading slash because the request
+               // could meant to list current__XL_DIR in the first place.
+               if err == errFileNotFound && strings.HasSuffix(current, slashSeparator) {
+                       current = strings.TrimSuffix(current, slashSeparator) + globalDirSuffixWithSlash
+                       entries, err = s.ListDir(ctx, opts.Bucket, current, -1)
+               }
                if err != nil {
                        // Folder could have gone away in-between
                        if err != errVolumeNotFound && err != errFileNotFound {
```

**Comment by klauspost** [comment]
@vadmeste Yeah, XLDIR hack bites again.

Thing is I am not even sure if there cannot be files with the same prefix, ie.

`prefix/`  <<<-- object
`prefix/file.txt` <<<---- object with same prefix.

I am highly unsure we even check that properly.

**Comment by klauspost** [comment]
TLDR; I think we may have to check for that on every root dir, even if we find a prefix. But check if it is possible to create this scenario.

**Comment by chrisharward** [comment]
Is this fix included in the latest release? (RELEASE.2021-04-06T23-11-00Z)

**Comment by harshavardhana** [comment]
> Is this fix included in the latest release? (RELEASE.2021-04-06T23-11-00Z)

This has been included since 02-07 here is how you check it 

```
git tag --contains c9b0f595b99a9c874b41c1399ebb57ac615c609b
RELEASE.2021-02-07T01-31-02Z
RELEASE.2021-02-11T08-23-43Z
RELEASE.2021-02-14T04-01-33Z
RELEASE.2021-02-19T04-38-02Z
RELEASE.2021-02-23T20-05-01Z
RELEASE.2021-02-24T18-44-45Z
RELEASE.2021-03-01T04-20-55Z
RELEASE.2021-03-04T00-53-13Z
RELEASE.2021-03-10T05-11-33Z
RELEASE.2021-03-12T00-00-47Z
RELEASE.2021-03-17T02-33-02Z
RELEASE.2021-03-26T00-00-41Z
RELEASE.2021-04-06T23-11-00Z
```

**Comment by chrisharward** [comment]
Thanks!
